plugins/jobs/src/js/components/JobFormModal.tsx

```diff
@@ -102,39 +145,251 @@ export default class JobFormModal extends React.Component<
     isEdit: false,
     job: new Job()
   };
+  private triggerFormSubmit: FormSubmitCallback;
 
   constructor() {
-    super(...arguments);
+    super();
 
     this.state = {
-      defaultTab: ""
+      jobFormModel: null,
+      jobJsonString: null,
+      isJsonMode: false
     };
 
-    this.handleTabChange = this.handleTabChange.bind(this);
+    this.triggerFormSubmit = undefined;
+
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleJSONEditorChange = this.handleJSONEditorChange.bind(this);
+    this.handleFormChange = this.handleFormChange.bind(this);
+    this.handleTriggerSubmit = this.handleTriggerSubmit.bind(this);
+    this.handleInputModeToggle = this.handleInputModeToggle.bind(this);
+  }
+
+  componentWillReceiveProps(nextProps: JobFormModalProps) {
+    if (nextProps.isOpen !== this.props.isOpen) {
+      this.resetState();
+    }
+  }
+
+  handleFormChange(event: { model: object }) {
+    if (!event.model) {
+      return;
+    }
+
+    this.props.handleErrorMessageChange(null);
+  }
+
+  handleSubmit() {
+    const { isEdit } = this.props;
+    const job = this.createJobFromEditorContents();
+    if (!job) {
+      return;
+    }
+
+    const jobSpec = JobUtil.createJobSpecFromJob(job);
+
+    this.props.handleSubmit(isEdit, jobSpec, job.getId());
+  }
+
+  handleInputModeToggle() {
+    const job = this.createJobFromEditorContents(true) || this.props.job;
+
+    if (this.state.isJsonMode) {
+      this.setState(
+        {
+          jobFormModel: JobUtil.createFormModelFromSchema(JobSchema, job),
+          isJsonMode: false
+        },
+        () => this.props.handleErrorMessageChange(null)
+      );
+    } else {
+      this.setState(
+        {
+          jobJsonString: JSON.stringify(
+            cleanJobJSON(JobUtil.createJobSpecFromJob(job as Job)),
+            null,
+            2
+          ),
+          isJsonMode: true
+        },
+        () => this.props.handleErrorMessageChange(null)
+      );
+    }
+  }
+
+  handleJSONEditorChange(jobJsonString: string) {
+    this.setState({ jobJsonString }, () =>
+      this.props.handleErrorMessageChange(null)
+    );
+  }
+
+  handleTriggerSubmit(submitFunction: () => DataTriple) {
+    this.triggerFormSubmit = submitFunction;
+  }
+
+  resetState() {
+    const { job } = this.props;
+    this.setState({
+      jobFormModel: JobUtil.createFormModelFromSchema(JobSchema, job),
+      jobJsonString: JSON.stringify(
+        cleanJobJSON(JobUtil.createJobSpecFromJob(job as Job)),
+        null,
+        2
+      )
+    });
+  }
+
+  createJobFromEditorContents(keepValidationErrors = false) {
```

Comment:
🎨 _Uff_…

Reply:
I appreciate your effort going through this piece of legacy, though I'd say we're keeping it for now.